File: baekjoon/3311_traffic.py
import sys
sys.setrecursionlimit(100000)
n, m, A, B = map(int, input().split())
Ve = [list(map(int, input().split())) for _ in range(n)]
E = [[] for _ in range(n)]
for _ in range(m):
    c, d, k = map(int, input().split())
    c -= 1
    d -= 1
    E[c].append(d)
    if k == 2:
        E[d].append(c)

LM = [i for i in range(n) if Ve[i][0] == 0]
RM = [i for i in range(n) if Ve[i][0] == A]

def scc(edge, n):
    from collections import deque
    grouped = [False for _ in range(n)]
    level_of = [False for _ in range(n)]
    stack = deque([])
    level = 0
    groups = []

    def dfs(u):
        nonlocal level
        level += 1
        last_level = level_of[u] = level

        stack.append(u)
        for v in edge[u]:
            if not level_of[v]:
                last_level = min(last_level, dfs(v))
            elif not grouped[v]:
                last_level = min(last_level, level_of[v])
        
        if last_level == level_of[u]:
            group = []
            while stack:
                p = stack.pop()
                group.append(p)
                grouped[p] = True
                if u == p:
                    break

            groups.append(group)
        return last_level

    for i in range(n):
        if not grouped[i]:
            dfs(i)
    
    return groups

groups = scc(E, n)
V_SCC = [None for _ in range(n)]
for i in range(len(groups)):
    for j in range(len(groups[i])):
        V_SCC[groups[i][j]] = i
groups_edge = {}
for i in range(len(groups)):
    groups_edge[i] = set()

for u in range(n):
    for v in E[u]:
        if V_SCC[u] != V_SCC[v]:
            groups_edge[V_SCC[u]].add(V_SCC[v])

SCC_E = [[] for _ in range(len(groups))]
for i in range(len(groups_edge)):
    SCC_E[i] = list(groups_edge[i])

V = [[False, float("inf"), -float("inf")] for _ in range(len(groups))]
for rm in RM:
    V[V_SCC[rm]][1] = min(V[V_SCC[rm]][1], Ve[rm][1])
    V[V_SCC[rm]][2] = max(V[V_SCC[rm]][2], Ve[rm][1])
SLM = sorted(LM, key=lambda x:Ve[x][1], reverse=True)
SRM = sorted(RM, key=lambda x:Ve[x][1], reverse=True)

SRMI = {}
for i in range(len(SRM)):
    SRMI[Ve[SRM[i]][1]] = len(SRM) - i
def dfs(u):
    if V[u][0]:
        return V[u][1], V[u][2]
    
    V[u][0] = True
    for v in SCC_E[u]:
        low, top = dfs(v)
        V[u][1] = min(V[u][1], low)
        V[u][2] = max(V[u][2], top)
    return V[u][1], V[u][2]

ans = []
for i in range(len(SLM)):
    low, top = dfs(V_SCC[SLM[i]])
    if low > top:
        ans.append(0)
    else:
        ans.append(SRMI[top] - SRMI[low] + 1)
print(*ans, sep="\n")

File: baekjoon/3311_traffic.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>
#include <map>
#include <climits>
#include <cmath>
#include <functional>
using namespace std;

int n, m, A, B;
vector<vector<int>> Ve;
vector<vector<int>> E;
vector<int> LM, RM;

vector<vector<int>> scc(const vector<vector<int>>& edge, int n) {
    vector<bool> grouped(n, false);
    vector<int> level_of(n, 0);
    stack<int> st;
    int level = 0;
    vector<vector<int>> groups;

    function<int(int)> dfs = [&](int u) -> int {
        level++;
        int last_level = level_of[u] = level;

        st.push(u);
        for (int v : edge[u]) {
            if (level_of[v] == 0) {
                last_level = min(last_level, dfs(v));
            } else if (!grouped[v]) {
                last_level = min(last_level, level_of[v]);
            }
        }

        if (last_level == level_of[u]) {
            vector<int> group;
            while (!st.empty()) {
                int p = st.top();
                st.pop();
                group.push_back(p);
                grouped[p] = true;
                if (u == p) break;
            }
            groups.push_back(group);
        }
        return last_level;
    };

    for (int i = 0; i < n; i++) {
        if (!grouped[i]) {
            dfs(i);
        }
    }

    return groups;
}

vector<int> V_SCC;
vector<vector<int>> SCC_E;
vector<vector<int>> V;
map<int, int> SRMI;
vector<bool> visited_group;

pair<int, int> dfs(int u) {
    if (V[u][0]) {
        return {V[u][1], V[u][2]};
    }
    
    V[u][0] = true;
    visited_group[u] = true;
    for (int v : SCC_E[u]) {
        pair<int, int> result = dfs(v);
        int low = result.first;
        int top = result.second;
        V[u][1] = min(V[u][1], low);
        V[u][2] = max(V[u][2], top);
    }
    return {V[u][1], V[u][2]};
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    cin >> n >> m >> A >> B;
    
    Ve.resize(n, vector<int>(2));
    for (int i = 0; i < n; i++) {
        cin >> Ve[i][0] >> Ve[i][1];
    }
    
    E.resize(n);
    for (int i = 0; i < m; i++) {
        int c, d, k;
        cin >> c >> d >> k;
        c--; d--;
        E[c].push_back(d);
        if (k == 2) {
            E[d].push_back(c);
        }
    }
    
    for (int i = 0; i < n; i++) {
        if (Ve[i][0] == 0) {
            LM.push_back(i);
        }
        if (Ve[i][0] == A) {
            RM.push_back(i);
        }
    }
    
    auto groups = scc(E, n);
    V_SCC.resize(n);
    
    for (int i = 0; i < groups.size(); i++) {
        for (int j = 0; j < groups[i].size(); j++) {
            V_SCC[groups[i][j]] = i;
        }
    }
    
    vector<vector<int>> groups_edge(groups.size());
    for (int u = 0; u < n; u++) {
        for (int v : E[u]) {
            if (V_SCC[u] != V_SCC[v]) {
                groups_edge[V_SCC[u]].push_back(V_SCC[v]);
            }
        }
    }
    
    SCC_E.resize(groups.size());
    visited_group.resize(groups.size(), false);
    for (int i = 0; i < groups.size(); i++) {
        // 중복 제거
        sort(groups_edge[i].begin(), groups_edge[i].end());
        groups_edge[i].erase(unique(groups_edge[i].begin(), groups_edge[i].end()), groups_edge[i].end());
        SCC_E[i] = groups_edge[i];
    }
    
    V.resize(groups.size(), vector<int>{0, INT_MAX, INT_MIN});
    for (int rm : RM) {
        V[V_SCC[rm]][1] = min(V[V_SCC[rm]][1], Ve[rm][1]);
        V[V_SCC[rm]][2] = max(V[V_SCC[rm]][2], Ve[rm][1]);
    }
    
    vector<pair<int, int>> SLM_pairs, SRM_pairs;
    for (int lm : LM) {
        SLM_pairs.push_back({Ve[lm][1], lm});
    }
    for (int rm : RM) {
        SRM_pairs.push_back({Ve[rm][1], rm});
    }
    
    sort(SLM_pairs.rbegin(), SLM_pairs.rend());
    sort(SRM_pairs.rbegin(), SRM_pairs.rend());
    
    vector<int> SLM, SRM;
    for (auto p : SLM_pairs) SLM.push_back(p.second);
    for (auto p : SRM_pairs) SRM.push_back(p.second);

    vector<vector<int>> ans;
    for (int i = 0; i < SLM.size(); i++) {
        pair<int, int> result = dfs(V_SCC[SLM[i]]);
        int low = result.first;
        int top = result.second;
        ans.push_back({low, top});
    }
    
    vector<int> remain_SRM;
    for (int srm : SRM){
        if (visited_group[V_SCC[srm]]){
            remain_SRM.push_back(srm);
        }
    }

    for (int i = 0; i < remain_SRM.size(); i++) {
        SRMI[Ve[remain_SRM[i]][1]] = remain_SRM.size() - i;
    }

    for (vector<int> result : ans) {
        if (result[1] < result[0]){
            cout << 0 << "\n";
        }
        else{
            cout << SRMI[result[1]] - SRMI[result[0]] + 1 << "\n";
        }
    }
    
    return 0;
}